### DISCUSSION TITLE
Weakly supervised metric learning?

### DISCUSSION THREAD

**Discussion by nikosmichas** [KevinMusgrave/pytorch-metric-learning#482] [Q&A] [answered]
Hello! Thanks for the great library.
I'm trying to find a way to use this library for weakly supervised task (same way as this does with scikit learn http://contrib.scikit-learn.org/metric-learn/weakly_supervised.html#learning-on-triplets)
Is it something that can be achieved using pytorch-metric-learning? Is there any tutorial / example for something similar?

**Comment by KevinMusgrave** [comment, marked as answer]
If you mean training on triplets alone instead of labels, then yes. Any tuple-based loss function takes in an argument `indices_tuple` which describes the pairs or triplets you want to train on.

For triplets, the format of `indices_tuple` is `(a, p, n)`, where `a`, `p`, and `n` are integer tensors:

```python
import torch
from pytorch_metric_learning.losses import TripletMarginLoss

loss_fn = TripletMarginLoss()
embeddings = torch.randn(32, 128)

# Two triplets. The first triplet has
# anchor: embeddings[0]
# positive: embeddings[1]
# negative: embeddings[5]
triplet_indices = [(0, 1, 5), (2, 3, 18)]

# "a" is all anchors, "p" is all positives, and "n" is all negatives
a, p, n = [torch.tensor(x) for x in list(zip(*triplet_indices))]

labels = torch.arange(len(embeddings)) # dummy labels
loss = loss_fn(embeddings, labels, indices_tuple = (a, p, n))
```

The dummy labels won't be used. I will make labels optional in a future version (see https://github.com/KevinMusgrave/pytorch-metric-learning/issues/412)

**Reply by nikosmichas** [reply]
@KevinMusgrave thanks for your reply.
Is there possibly a complete example with a similar usage? 
Thanks in advance

**Reply by KevinMusgrave** [reply]
No unfortunately

**Reply by KevinMusgrave** [reply]
Starting in v1.5.0, losses like ContrastiveLoss and TripletMarginLoss no longer require labels if `indices_tuple` is passed in:

```python
loss = loss_fn(embeddings, indices_tuple=triplets)
```

(Posting here for future readers.)
